File: src/dotnet/Core/Application/Services/CandidateAggregator.cs
using Rag.Candidates.Core.Application.DTOs;

namespace Rag.Candidates.Core.Application.Services;

public sealed class CandidateAggregator
{
    private readonly string _candidateIdField;

    public CandidateAggregator(string candidateIdField)
    {
        _candidateIdField = candidateIdField;
    }

    public List<AggregatedCandidate> Aggregate(List<(string Document, Dictionary<string, object> Metadata, double Score)> searchResults)
    {
        var candidateMap = new Dictionary<string, CandidateData>();

        foreach (var (document, metadata, score) in searchResults)
        {
            var candidateId = metadata.TryGetValue(_candidateIdField, out var id) 
                ? id.ToString() ?? "unknown" 
                : "unknown";

            if (!candidateMap.ContainsKey(candidateId))
            {
                candidateMap[candidateId] = new CandidateData
                {
                    CandidateId = candidateId,
                    Documents = new List<string>(),
                    Metadata = new Dictionary<string, object>(metadata),
                    Scores = new List<double>()
                };
            }

            candidateMap[candidateId].Documents.Add(document);
            candidateMap[candidateId].Scores.Add(score);
        }

        var aggregated = new List<AggregatedCandidate>();
        foreach (var candidateData in candidateMap.Values)
        {
            aggregated.Add(new AggregatedCandidate
            {
                CandidateId = candidateData.CandidateId,
                Documents = candidateData.Documents,
                Metadata = candidateData.Metadata,
                MaxScore = candidateData.Scores.Count > 0 ? candidateData.Scores.Max() : 0.0,
                AllScores = candidateData.Scores
            });
        }

        return aggregated;
    }

    private class CandidateData
    {
        public required string CandidateId { get; init; }
        public required List<string> Documents { get; init; }
        public required Dictionary<string, object> Metadata { get; init; }
        public required List<double> Scores { get; init; }
    }
}

File: src/python/core/application/services/candidate_aggregator.py
from typing import List, Dict, Any
from dataclasses import dataclass


@dataclass
class AggregatedCandidate:
    candidate_id: str
    documents: List[str]
    metadata: Dict[str, Any]
    max_score: float
    all_scores: List[float]


class CandidateAggregator:
    def __init__(self, candidate_id_field: str):
        self._candidate_id_field = candidate_id_field
    
    def aggregate(self, search_results: List[tuple]) -> List[AggregatedCandidate]:
        candidate_map = {}
        
        for document, metadata, score in search_results:
            candidate_id = metadata.get(self._candidate_id_field, "unknown")
            
            if candidate_id not in candidate_map:
                candidate_map[candidate_id] = {
                    "candidate_id": candidate_id,
                    "documents": [],
                    "metadata": metadata.copy(),
                    "scores": []
                }
            
            candidate_map[candidate_id]["documents"].append(document)
            candidate_map[candidate_id]["scores"].append(score)
        
        aggregated = []
        for candidate_data in candidate_map.values():
            aggregated.append(AggregatedCandidate(
                candidate_id=candidate_data["candidate_id"],
                documents=candidate_data["documents"],
                metadata=candidate_data["metadata"],
                max_score=max(candidate_data["scores"]) if candidate_data["scores"] else 0.0,
                all_scores=candidate_data["scores"]
            ))
        
        return aggregated
